File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.js
const offsets = [
  [-1, 0], // up
  [1, 0], // down
  [0, -1], // left
  [0, 1], // right
];

/**
 * @param {number[][]} heights
 * @return {number[][]}
 */
var pacificAtlantic = function (heights) {
  const m = heights.length;
  const n = heights[0].length;

  const isValid = (r, c) => 0 <= r && 0 <= c && r < m && c < n;

  const peaks = Array(m)
    .fill(null)
    .map(() => Array(n).fill(0));

  const res = [];

  const findPeak = (r, c, ocean, prevHeight) => {
    if (
      !isValid(r, c) ||
      peaks[r][c] === ocean ||
      peaks[r][c] === 3 ||
      heights[r][c] < prevHeight
    ) {
      return;
    }

    peaks[r][c] += ocean;
    if (peaks[r][c] === 3) {
      res.push([r, c]);
    }

    for (const [dr, dc] of offsets) {
      findPeak(r + dr, c + dc, ocean, heights[r][c]);
    }
  };

  const pacific = 1;
  const atlantic = 2;

  for (let r = 0; r < m; r++) {
    findPeak(r, 0, pacific, heights[r][0]);
    findPeak(r, n - 1, atlantic, heights[r][n - 1]);
  }
  for (let c = 0; c < n; c++) {
    findPeak(0, c, pacific, heights[0][c]);
    findPeak(m - 1, c, atlantic, heights[m - 1][c]);
  }

  return res;
};

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.go
func pacificAtlantic(heights [][]int) (coords [][]int) {
    m, n := len(heights), len(heights[0])
  
    isValid := func (r, c int) bool {
        return r >= 0 && c >=0 && r < m && c < n
    }
 
    peaks := make([]int, m * n)
  
    var findPeak func (r, c, ocean, prevHeight int)
  
    findPeak = func (r, c, ocean, prevHeight int) {
        i := r * n + c
    
        if !isValid(r, c) || peaks[i] == ocean || peaks[i] == 3 || heights[r][c] < prevHeight {
            return
        }
    
        peaks[i] += ocean;
        if peaks[i] == 3 {
            coords = append(coords, []int{r, c})
        }

        findPeak(r - 1, c, ocean, heights[r][c])
        findPeak(r + 1, c, ocean, heights[r][c])
        findPeak(r, c - 1, ocean, heights[r][c])
        findPeak(r, c + 1, ocean, heights[r][c])
    }
  
    pacific, atlantic := 1, 2

    for r := 0; r < m; r++ {
        findPeak(r, 0, pacific, heights[r][0])
        findPeak(r, n - 1, atlantic, heights[r][n - 1])
    }
    for c := 0; c < n; c++ {
        findPeak(0, c, pacific, heights[0][c])
        findPeak(m - 1, c, atlantic, heights[m - 1][c])
    }

    return
}